// @hono/vite-dev-server/dist/dev-server.js
import { getRequestListener } from "@hono/node-server";
import { minimatch } from "minimatch";
import fs from "fs";
import path from "path";
import { createBasePathGuard, createBasePathRewriter, safeParseUrlPath } from "./utils.js";
const defaultOptions = {
  entry: "./src/index.ts",
  export: "default",
  injectClientScript: true,
  exclude: [
    /.*\.css$/,
    /.*\.ts$/,
    /.*\.tsx$/,
    /.*\.mdx?$/,
    /^\/@.+$/,
    /\?t\=\d+$/,
    /[?&]tsr-split=[^&]*(&t=[^&]*)?$/,
    // Support for TanStack Router code splitting
    /^\/favicon\.ico$/,
    /^\/static\/.+/,
    /^\/node_modules\/.*/,
    /.*\.svelte$/,
    /.*\.vue$/,
    /.*\.js$/,
    /.*\.jsx$/
  ],
  ignoreWatching: [/\.wrangler/, /\.mf/],
  handleHotUpdate: ({ server, modules }) => {
    const isSSR = modules.some((mod) => mod._ssrModule);
    if (isSSR) {
      server.hot.send({ type: "full-reload" });
      return [];
    }
  },
  base: ""
};
const defaultBase = "/";
function devServer(options) {
  let publicDirPath = "";
  let viteBase = defaultBase;
  const entry = options?.entry ?? defaultOptions.entry;
  const rewriteRequestForBase = createBasePathRewriter(options?.base ?? defaultOptions.base);
  const shouldHandlePath = createBasePathGuard(options?.base ?? defaultOptions.base);
  let viteBaseSetByUser;
  const plugin = {
    name: "@hono/vite-dev-server",
    config: (config) => {
      viteBaseSetByUser = config?.base;
      const baseOption = options?.base;
      return {
        ...baseOption !== void 0 ? {
          base: baseOption === "" ? "/" : baseOption
        } : {},
        server: {
          watch: {
            ignored: options?.ignoreWatching ?? defaultOptions.ignoreWatching
          }
        }
      };
    },
    configResolved(config) {
      publicDirPath = config.publicDir;
      viteBase = config.base;
      const honoBase = options?.base;
      const honoBaseIsSet = honoBase !== void 0;
      const viteBaseIsSet = viteBaseSetByUser !== void 0 && viteBaseSetByUser !== "/";
      if (viteBaseIsSet && honoBaseIsSet) {
        console.warn(
          `The following Vite config options will be overridden by @hono/vite-dev-server:
  - base: "${viteBaseSetByUser}" -> "${honoBase}"`
        );
      }
    },
    configureServer: async (server) => {
      async function createMiddleware(server2) {
        return async function(req, res, next) {
          if (req.url) {
            const filePath = path.join(publicDirPath, req.url);
            try {
              if (fs.existsSync(filePath) && fs.statSync(filePath).isFile()) {
                return next();
              }
            } catch {
            }
          }
          const exclude = options?.exclude ?? defaultOptions.exclude;
          for (const pattern of exclude) {
            if (req.url) {
              if (pattern instanceof RegExp) {
                if (pattern.test(req.url)) {
                  return next();
                }
              } else if (minimatch(req.url?.toString(), pattern)) {
                return next();
              }
            }
          }
          const pathname = req.url ? safeParseUrlPath(req.url) : void 0;
          if (!shouldHandlePath(pathname)) {
            return next();
          }
          let loadModule;
          if (options?.loadModule) {
            loadModule = options.loadModule;
          } else {
            loadModule = async (server3, entry2) => {
              let appModule;
              try {
                appModule = await server3.ssrLoadModule(entry2);
              } catch (e) {
                if (e instanceof Error) {
                  server3.ssrFixStacktrace(e);
                }
                throw e;
              }
              const exportName = options?.export ?? defaultOptions.export;
              const app2 = appModule[exportName];
              if (!app2) {
                throw new Error(`Failed to find a named export "${exportName}" from ${entry2}`);
              }
              return app2;
            };
          }
          let app;
          try {
            app = await loadModule(server2, entry);
          } catch (e) {
            return next(e);
          }
          getRequestListener(
            async (request) => {
              let env = {
                incoming: req,
                outgoing: res,
                vite: server2
              };
              if (options?.env) {
                if (typeof options.env === "function") {
                  env = { ...env, ...await options.env() };
                } else {
                  env = { ...env, ...options.env };
                }
              }
              const adapter = await getAdapterFromOptions(options);
              if (adapter?.env) {
                env = { ...env, ...adapter.env };
              }
              const executionContext = adapter?.executionContext ?? {
                waitUntil: async (fn) => fn,
                passThroughOnException: () => {
                  throw new Error("`passThroughOnException` is not supported");
                }
              };
              const requestForApp = rewriteRequestForBase ? rewriteRequestForBase(request) : request;
              const response = await app.fetch(requestForApp, env, executionContext);
              if (!(response instanceof Response)) {
                throw response;
              }
              if (options?.injectClientScript !== false && response.headers.get("content-type")?.match(/^text\/html/)) {
                const viteScript = path.posix.join(viteBase, "/@vite/client");
                const nonce = response.headers.get("content-security-policy")?.match(/'nonce-([^']+)'/)?.[1];
                const script = `<script${nonce ? ` nonce="${nonce}"` : ""}>import("${viteScript}")</script>`;
                return injectStringToResponse(response, script) ?? response;
              }
              return response;
            },
            {
              overrideGlobalObjects: false,
              errorHandler: (e) => {
                let err;
                if (e instanceof Error) {
                  err = e;
                  server2.ssrFixStacktrace(err);
                } else if (typeof e === "string") {
                  err = new Error(`The response is not an instance of "Response", but: ${e}`);
                } else {
                  err = new Error(`Unknown error: ${e}`);
                }
                next(err);
              }
            }
          )(req, res);
        };
      }
      server.middlewares.use(await createMiddleware(server));
      server.httpServer?.on("close", async () => {
        const adapter = await getAdapterFromOptions(options);
        if (adapter?.onServerClose) {
          await adapter.onServerClose();
        }
      });
    },
    handleHotUpdate: options?.handleHotUpdate ?? defaultOptions.handleHotUpdate
  };
  return plugin;
}
const getAdapterFromOptions = async (options) => {
  let adapter = options?.adapter;
  if (typeof adapter === "function") {
    adapter = adapter();
  }
  if (adapter instanceof Promise) {
    adapter = await adapter;
  }
  return adapter;
};
function injectStringToResponse(response, content) {
  const stream = response.body;
  const newContent = new TextEncoder().encode(content);
  if (!stream) {
    return null;
  }
  const reader = stream.getReader();
  const combinedStream = new ReadableStream({
    async start(controller) {
      for (; ; ) {
        const result = await reader.read();
        if (result.done) {
          break;
        }
        controller.enqueue(result.value);
      }
      controller.enqueue(newContent);
      controller.close();
    }
  });
  const headers = new Headers(response.headers);
  headers.delete("content-length");
  return new Response(combinedStream, {
    headers,
    status: response.status
  });
}
export {
  defaultOptions,
  devServer
};
