// react-router-hono-server/dist/adapters/bun.js
import {
  bindIncomingRequestSocketInfo,
  cleanUpgradeListeners,
  createGetLoadContext,
  createWebSocket,
  getBuildMode,
  importBuild,
  patchUpgradeListener
} from "../chunk-BH64OQMW.js";
import {
  cache
} from "../chunk-YI5FED4X.js";

// src/adapters/bun.ts
import { Hono } from "hono";
import { serveStatic } from "hono/bun";
import { createMiddleware } from "hono/factory";
import { logger } from "hono/logger";
import { createRequestHandler } from "react-router";
var serverInstance;
var shutdownCallback;
async function createHonoServer(options) {
  const build = await importBuild();
  const basename = import.meta.env.REACT_ROUTER_HONO_SERVER_BASENAME;
  const mergedOptions = {
    ...options,
    port: options?.port || Number(options?.customBunServer?.port) || Number(process.env.PORT) || 3e3,
    defaultLogger: options?.defaultLogger ?? true
  };
  const mode = getBuildMode();
  const PRODUCTION = mode === "production";
  const clientBuildPath = `${import.meta.env.REACT_ROUTER_HONO_SERVER_BUILD_DIRECTORY}/client`;
  if (PRODUCTION && mergedOptions.onGracefulShutdown) {
    shutdownCallback = mergedOptions.onGracefulShutdown;
  }
  const app = new Hono(mergedOptions.app);
  const { upgradeWebSocket, injectWebSocket } = await createWebSocket({
    app,
    enabled: mergedOptions.useWebSocket ?? false
  });
  if (!PRODUCTION) {
    app.use(bindIncomingRequestSocketInfo());
  }
  await mergedOptions.beforeAll?.(app);
  app.use(
    `/${import.meta.env.REACT_ROUTER_HONO_SERVER_ASSETS_DIR}/*`,
    cache(60 * 60 * 24 * 365),
    // 1 year
    serveStatic({
      root: clientBuildPath,
      ...mergedOptions.serveStaticOptions?.clientAssets
    })
  );
  app.use(
    "*",
    cache(60 * 60),
    // 1 hour
    serveStatic({
      root: PRODUCTION ? clientBuildPath : "./public",
      ...mergedOptions.serveStaticOptions?.publicAssets
    })
  );
  if (mergedOptions.defaultLogger) {
    app.use("*", logger());
  }
  if (mergedOptions.useWebSocket) {
    await mergedOptions.configure(app, { upgradeWebSocket });
  } else {
    await mergedOptions.configure?.(app);
  }
  const reactRouterApp = new Hono({
    strict: false
  });
  reactRouterApp.use((c, next) => {
    return createMiddleware(async (c2) => {
      const requestHandler = createRequestHandler(build, mode);
      const loadContext = mergedOptions.getLoadContext?.(c2, { build, mode });
      return requestHandler(c2.req.raw, loadContext instanceof Promise ? await loadContext : loadContext);
    })(c, next);
  });
  app.route(`${basename}`, reactRouterApp);
  if (basename) {
    app.route(`${basename}.data`, reactRouterApp);
  }
  let server = {
    ...mergedOptions.customBunServer,
    fetch: app.fetch,
    port: mergedOptions.port,
    development: !PRODUCTION
  };
  if (PRODUCTION) {
    server = injectWebSocket(server);
    serverInstance = Bun.serve(server);
    console.log(`Started server: ${serverInstance.protocol}://${serverInstance.hostname}:${serverInstance.port}`);
    if (shutdownCallback != null) {
      let gracefulShutdownInProgress = false;
      const gracefulShutdown = async (signal) => {
        if (gracefulShutdownInProgress) return;
        gracefulShutdownInProgress = true;
        try {
          if (serverInstance != null) {
            console.log(`
\u{1F4E1} Received ${signal}, shutting down gracefully...`);
            await serverInstance.stop(false);
            console.log("Server stopped, all requests completed");
            if (shutdownCallback) {
              console.log("\u{1F9F9} Running cleanup tasks...");
              await shutdownCallback();
              console.log("\u2705 Cleanup complete!");
            }
            console.log("\u{1F44B} Graceful shutdown completed");
          }
          process.exit(0);
        } catch (error) {
          console.error("\u274C Shutdown failed:", error);
          process.exit(1);
        }
      };
      if (serverInstance != null) {
        process.on("SIGTERM", () => gracefulShutdown("SIGTERM"));
        process.on("SIGINT", () => gracefulShutdown("SIGINT"));
        console.log("\u2705 Graceful shutdown enabled. Press Ctrl+C to shutdown gracefully.");
      }
    }
  } else if (globalThis.__viteDevServer?.httpServer) {
    const httpServer = globalThis.__viteDevServer.httpServer;
    cleanUpgradeListeners(httpServer);
    injectWebSocket(httpServer);
    patchUpgradeListener(httpServer);
    console.log("\u{1F6A7} Running in development mode");
  }
  return server;
}
export {
  createGetLoadContext,
  createHonoServer
};
